File: examples/account_sample.pl
#!/usr/bin/perl

=pod

=head1 NAME

account_sample.pl - Simple example of using the Open Hub API from Perl

=head1 SYNOPSIS

account_sample.pl api_key email

=head1 DESCRIPTION

This is a small example of how to use the Open Hub API from Perl,
mimicking the Ruby example - account_sample.rb

Detailed information about the Open Hub API can be found at
https://github.com/blackducksw/ohloh_api

The script takes an API key and the email address of a developer
registered on Open Hub as parameters, and prints out all the
information of that account.

=head1 EXAMPLE

    ./account_sample.pl YOUR_API_KEY_HERE robin@ohloh.net

=head1 DEPENDENCIES

This script uses
L<LWP::Simple> (to query the Open Hub website),
L<Digest::MD5> (to MD5s the email address as requested by the API)
and
L<XML::Simple> (to parse the server's answer).

=head1 VERSION

This documentation refers to account_sample version 0.1

=head1 AUTHOR

Yanick Champoux (yanick@cpan.org)

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007 Yanick Champoux (yanick@cpan.org). All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perldoc perlartistic.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

use strict;
use warnings;

use LWP::Simple;
use Digest::MD5 qw/ md5_hex /;
use XML::Simple;

my $VERSION     = '0.01';
my $API_VERSION = 1;

die_usage() unless @ARGV == 2;

my ( $api_key, $email ) = @ARGV;

my $email_md5 = md5_hex($email);

### let's build the request url
# first the base url
my $url = 'https://www.openhub.net/';
# then the email "page" we want
$url .= "accounts/$email_md5.xml";
# and finally the API version and the api key
$url .= '?' . join '&' => "v=$API_VERSION", "api_key=$api_key";

# request the url from the server
my $response = get $url or die "Open Hub server didn't return anything\n";

# parse the XML response
my $xml = eval { XMLin($response) } or die "Server didn't return valid XML\n";

# was the request a success?
die "request didn't succeed: $xml->{error}\n"
  unless $xml->{status} eq 'success';

# make an alias to the info we want
my %account = %{ $xml->{result}{account} };

# now we print the info for the account

for my $attr (
    qw/ id name created_at updated_at homepage_url
    avatar_url posts_count location country_code
    latitude longitude /
  ) {

    # only print the attribute if it exist
    printf "%-20s: %s\n", $attr, $account{$attr}
      if defined $account{$attr} and not ref $account{$attr};
}

# print the kudo information
printf "%-20s: %s\n", 'kudo rank', $account{kudo_score}{kudo_rank}
  if $account{kudo_score};

exit;

### utility functions ######################################

sub die_usage {
    die "usage: $0 api_key email\n";
}

__END__

File: examples/account_sample.rb
#!/usr/bin/env ruby
require 'net/https'
require 'rexml/document'
require 'digest/md5'

# The MIT License (MIT)
# 
# Copyright (c) 2013 Black Duck Software, Inc.
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

# This is an example of using the Open Hub API from Ruby.
# Detailed information can be found at the Open Hub website:
#
#     https://github.com/blackducksw/ohloh_api
#
# This example uses the REXML library for XML parsing:
#
#     http://www.germane-software.com/software/rexml/
#

# This example retrieves basic Open Hub account information
# and outputs it as simple name: value pairs.
#
# Pass your Open Hub API key as the first parameter to this script.
# Open Hub API keys are free. If you do not have one, you can obtain one
# at the Open Hub website:
#
#     https://www.openhub.net/accounts/<your_login>/api_keys/new
#
# Pass the email address of the account as the second parameter to this script.

unless ARGV[0] =~ /[\w]+/ and ARGV[1] =~ /.+@.+\..+/
	STDERR.puts "Usage: #{__FILE__} [api_key] [email]"
	exit 1
end

api_key = ARGV[0]
email = ARGV[1]

#
# Connect to the Open Hub website and retrieve the account data.
#

# We pass the MD5 hash of the email address
email_md5 = Digest::MD5.hexdigest(email).to_s

uri = URI("https://www.openhub.net/accounts/#{email_md5}.xml?v=1&api_key=#{api_key}")
http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true
response = http.get(uri.request_uri)

# HTTP OK?
if response.code != '200'
  STDERR.puts "#{response.code} - #{response.message}"
  exit 1
end

# Parse the response into a structured XML object
xml = REXML::Document.new(response.body)

# Did Open Hub return an error?
error = xml.root.get_elements('/response/error').first
if error
  STDERR.puts "#{error.text}"
  exit 1
end

# Output all the immediate child properties of an Account
xml.root.get_elements('/response/result/account').first.each_element do |element|
  puts "#{element.name}:\t#{element.text}" unless element.has_elements?
end

# The kudo_rank element may not be present.
# Check for it, and default to 1 if it is missing.
xml_kudo_rank = xml.root.get_elements('/response/result/account/kudo_score/kudo_rank').first
kudo_rank = xml_kudo_rank ? xml_kudo_rank.text.to_i : 1
puts "kudo_rank:\t#{kudo_rank}"
